config/bidderinfo.go

```diff
@@ -177,94 +166,345 @@ type SyncerEndpoint struct {
-func LoadBidderInfoFromDisk(path string, adapterConfigs map[string]Adapter, bidders []string) (BidderInfos, error) {
-	reader := infoReaderFromDisk{path}
-	return loadBidderInfo(reader, adapterConfigs, bidders)
+func validateMaintainer(info *MaintainerInfo, bidderName string) error {
+	if info == nil || info.Email == "" {
+		return fmt.Errorf("missing required field: maintainer.email for adapter: %s", bidderName)
+	}
+	return nil
 }
 
-func loadBidderInfo(r infoReader, adapterConfigs map[string]Adapter, bidders []string) (BidderInfos, error) {
-	infos := BidderInfos{}
+func validateCapabilities(info *CapabilitiesInfo, bidderName string) error {
+	if info == nil {
+		return fmt.Errorf("missing required field: capabilities for adapter: %s", bidderName)
+	}
 
-	for _, bidder := range bidders {
-		data, err := r.Read(bidder)
-		if err != nil {
-			return nil, err
+	if info.App == nil && info.Site == nil {
+		return fmt.Errorf("at least one of capabilities.site or capabilities.app must exist for adapter: %s", bidderName)
+	}
+
+	if info.App != nil {
+		if err := validatePlatformInfo(info.App); err != nil {
+			return fmt.Errorf("capabilities.app failed validation: %v for adapter: %s", err, bidderName)
 		}
+	}
 
-		info := BidderInfo{}
-		if err := yaml.Unmarshal(data, &info); err != nil {
-			return nil, fmt.Errorf("error parsing yaml for bidder %s: %v", bidder, err)
+	if info.Site != nil {
+		if err := validatePlatformInfo(info.Site); err != nil {
+			return fmt.Errorf("capabilities.site failed validation: %v, for adapter: %s", err, bidderName)
 		}
+	}
+	return nil
+}
 
-		info.Enabled = isEnabledByConfig(adapterConfigs, bidder)
-		infos[bidder] = info
+func validatePlatformInfo(info *PlatformInfo) error {
+	if len(info.MediaTypes) == 0 {
+		return errors.New("mediaTypes should be an array with at least one string element")
 	}
 
-	return infos, nil
-}
+	for index, mediaType := range info.MediaTypes {
+		if mediaType != "banner" && mediaType != "video" && mediaType != "native" && mediaType != "audio" {
+			return fmt.Errorf("unrecognized media type at index %d: %s", index, mediaType)
+		}
+	}
 
-func isEnabledByConfig(adapterConfigs map[string]Adapter, bidderName string) bool {
-	a, ok := adapterConfigs[strings.ToLower(bidderName)]
-	return ok && !a.Disabled
+	return nil
 }
 
-type infoReader interface {
-	Read(bidder string) ([]byte, error)
+func validateSyncer(bidderInfo BidderInfo) error {
+	if bidderInfo.Syncer == nil {
+		return nil
+	}
+
+	for _, supports := range bidderInfo.Syncer.Supports {
+		if !strings.EqualFold(supports, "iframe") && !strings.EqualFold(supports, "redirect") {
+			return fmt.Errorf("syncer could not be created, invalid supported endpoint: %s", supports)
+		}
+	}
+
+	return nil
 }
 
-type infoReaderFromDisk struct {
-	path string
+func applyBidderInfoConfigOverrides(configBidderInfos BidderInfos, fsBidderInfos BidderInfos) (BidderInfos, error) {
```

Comment:
This is no longer necessary. We can let Viper take care of this work. You need to only give Viper the bidder infos before calling unmarshal and it will take care of the overrides:

```
func New(v *viper.Viper, bidderInfos BidderInfos) (*Configuration, error) {
	var c Configuration
	c.BidderInfos = bidderInfos

	if err := v.Unmarshal(&c); err != nil {
		return nil, fmt.Errorf("viper failed to unmarshal app config: %v", err)
	}
	
	...
```

This breaks is the legacy user sync url, which we can either choose to remove as part of this work or we alert `applyBidderInfoConfigOverrides` to only apply the legacy user sync url override. 

This also removes our ability to disallow certain parts of the config to be changed, like the maintainer or media types. The later can cause a problem if it's less restrictive.

I think we can either:
1. Let Viper handle everything and remove the override code. This allows hosts to "shoot themselves in the foot", but there's nothing stop them from editing the bidder info files directly?
2. Keep control of overrides and create a new BidderInfoOverride with less fields and leave the override code.

Thoughts?

Reply:
This is what you showed me on our very first discussion about the approach to support env variables. 
We also had a discussion about it with Hans: https://github.com/prebid/prebid-server/pull/2353#discussion_r963124663

If I set `c.BidderInfos = bidderInfos` and have env variable for at least one field for a bidder (for example endpoint) - it overrides the entire object for this bidder and sets "endpoint" only. Other data from yaml file is not present for this bidder after config umnarshalling. 

I think we need to control overrides. I removed `capabilities` from overrides, we can discuss what other fields should be removed. But I also think this is not obvious why some of the fields can be overridden and some not. Please advice.